File: identicon/src/main/java/com/lelloman/identicon/IdenticonView.java
package com.lelloman.identicon;

import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.widget.ImageView;

/**
 * an ImageView that displays an IdenticonDrawable, the type of
 * the IdenticonDrawable is unknown and the method to create it
 * is abstract, the alternative would be to pass the drawable directly
 * to the view but setting just the hash looks nicer
 */
public abstract class IdenticonView extends ImageView {


	private int mHash;
	private IdenticonDrawable mIdenticonDrawable;

	public IdenticonView(Context context) {
		super(context);
	}

	public IdenticonView(Context context, AttributeSet attrs) {
		super(context, attrs);
	}


	@Override
	protected void onSizeChanged(int w, int h, int oldw, int oldh) {
		super.onSizeChanged(w, h, oldw, oldh);

		if (w > 0 && h > 0) {
			mIdenticonDrawable = makeIdenticonDrawable(getWidth(), getHeight(), mHash);
			// TODO on Oreo the drawable are very stretched, for some reason ?_?
			if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O){
				mIdenticonDrawable.invalidateBitmap();
				setImageBitmap(mIdenticonDrawable.getBitmap());
			}else {
				setImageDrawable(mIdenticonDrawable);
			}
		}
	}

	public void setHash(int hash) {
		mHash = hash;
		if(mIdenticonDrawable != null) {
			mIdenticonDrawable.setHash(mHash);
			invalidate();
		}
	}

	protected abstract IdenticonDrawable makeIdenticonDrawable(int width, int height, int hash);
}

File: identicon/src/main/java/com/lelloman/identicon/view/IdenticonView.kt
package com.lelloman.identicon.view

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import androidx.appcompat.widget.AppCompatImageView
import com.lelloman.identicon.R
import com.lelloman.identicon.drawable.IdenticonDrawable
import com.lelloman.identicon.util.toIdenticonHash

/**
 * An [AppCompatImageView] that displays an [IdenticonDrawable], its instantiation is delegated to
 * [IdenticonView.makeIdenticonDrawable].
 */
abstract class IdenticonView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : AppCompatImageView(context, attrs) {

    var hash: ByteArray = ByteArray(16)
        set(value) {
            field = value
            identiconDrawable?.let { identiconDrawable ->
                identiconDrawable.hash = value
                invalidate()
            }
        }
    private var identiconDrawable: IdenticonDrawable? = null

    init {
        if (attrs != null) {
            val a = context.obtainStyledAttributes(attrs, R.styleable.IdenticonView)
            try {
                val intHash = a.getInt(R.styleable.IdenticonView_hash, 0)
                hash = intHash.toIdenticonHash()
            } catch (exception: Throwable) {
                Log.w(javaClass.simpleName, "Something went wrong when initializing ${javaClass.name} $this")
            } finally {
                a.recycle()
            }
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (w > 0 && h > 0) {
            identiconDrawable = makeIdenticonDrawable(width, height, hash)
            setImageDrawable(identiconDrawable)
        }
    }

    protected abstract fun makeIdenticonDrawable(width: Int, height: Int, hash: ByteArray): IdenticonDrawable
}
